### PR TITLE
Fix more bugs

### PR DESCRIPTION
**Pull request by mariane-bekker**

### CODE REVIEW

File: lib/RequestRouter.js [hunk 1]
```diff
@@ -336,7 +337,7 @@ function validateAndCastObject(object, schema, parentKey) {
             for (var arrayKey in value) {
                 arrayValue = validateAndCastParam.call(this, key, value[arrayKey], schemaProperty.items);
                 if (arrayValue && arrayValue.errorCode) {
-                    this.addPropertyError(parentKey + key, arrayValue.errorCode);
+                    this.addPropertyError(parentKey + key + '.' + arrayKey, arrayValue.errorCode);
```
Comment (mariane-bekker) [inline lib/RequestRouter.js]: @djvirgen Should we return for example the array key on error property: paramArray.0 in the error object or just property: paramArray?

Comment (djvirgen) [inline lib/RequestRouter.js]: the property error should be in the monocle-props format, e.g. `myObj.myProperty@myPluckedKey`

File: lib/RequestRouter.js [hunk 2]
```diff
@@ -442,7 +442,7 @@ RequestRouter.prototype.setResource = function(resource) {
 
     var schema = this._schema;
 
-    if (this._request.getMethod() === Request.METHOD_POST && this.isCollection()) {
+    if (!!~[Request.METHOD_PATCH, Request.METHOD_POST, Request.METHOD_PUT, Request.METHOD_DELETE].indexOf(this.getMethod()) && this.isCollection()) {
```
Comment (djvirgen) [inline lib/RequestRouter.js]: This should only be for `POST`.

File: lib/RequestRouter.js [hunk 3]
```diff
@@ -496,8 +496,14 @@ RequestRouter.prototype.addPropertyError = function(property, errorCode) {
         }
 
     }
+
+    //Remove the index of the array if it's added to the end of the property name
+    //so we can find the parent property in the schema
+    //Ex: paramArray.0 will become paramArray
+    var propertyTemp = _.clone(property).replace(/(\.[0-9]+)$/, '');
```
Comment (djvirgen) [inline lib/RequestRouter.js]: If possible, I suggest sticking with the monocle-props syntax. Check out the monocle-props module for useful utility methods like `.set()` and `.get()`

File: lib/RequestRouter.js [hunk 4]
```diff
@@ -514,8 +520,8 @@ RequestRouter.prototype.addPropertyError = function(property, errorCode) {
         });
     }
 
-    if (this._request.getMethod() == Request.METHOD_POST && this.isCollection()) {
-        //remove the appended items@ if posting to a collection
+    if (!!~[Request.METHOD_PATCH, Request.METHOD_POST, Request.METHOD_PUT, Request.METHOD_DELETE].indexOf(this.getMethod()) && this.isCollection()) {
```
Comment (djvirgen) [inline lib/RequestRouter.js]: This should only be for `POST`

File: test/lib/Router_test.js [hunk 5]
```diff
@@ -1248,7 +1266,7 @@ describe('API Router', function() {
                 var item = {
                     param4: ['hey']
                 };
-                return this.connection.post('/postit', {
+                return this.connection.patch('/postit', {
```
Comment (djvirgen) [inline test/lib/Router_test.js]: `.post`

File: lib/RequestRouter.js (hunk 4 above)
Comment (mariane-bekker) [inline lib/RequestRouter.js]: We also want to patch and put stuff for collections as well... we actually have a PATCH use case alreayd in our code that uses that

Comment (djvirgen) [inline lib/RequestRouter.js]: If you want to patch an individual item in a collection, you would use the uri of that specific item. For example, to create a new User I would `POST` a single object to `/users`, which is normally a collection endpoint. To update the user, we would make a `PATCH` request to that object's URL, e.g. `/users/123`, which is not a collection endpoint.

It really comes down to `POST` not being idempotent, while the other methods are. So we need to handle it differently.

File: lib/RequestRouter.js (hunk 1 above)
Comment (mariane-bekker) [inline lib/RequestRouter.js]: sounds good! will update

File: lib/RequestRouter.js (hunk 4 above)
Comment (djvirgen) [inline lib/RequestRouter.js]: When patching a collection, we're expecting to change the properties of the collection as a whole. Not individual items in the collection. I mean, which item would be modified? If I `PATCH /users`, which user gets edited? That wouldn't make sense..

File: lib/RequestRouter.js [hunk 6]
```diff
@@ -514,8 +518,8 @@ RequestRouter.prototype.addPropertyError = function(property, errorCode) {
         });
     }
 
-    if (this._request.getMethod() == Request.METHOD_POST && this.isCollection()) {
-        //remove the appended items@ if posting to a collection
+    if (this.isPost() && this.isCollection()) {
+        //remove the appended items@ if posting/patching/putting in a collection
```
Comment (djvirgen) [inline lib/RequestRouter.js]: This comment is now incorrect..

**Comment by djvirgen** [comment]
:cherries: once the comment is fixed

**Pull request merged**